### PR TITLE
Prepared ScheduleTask for threading refactor

### PR DESCRIPTION
**Pull request by tigerw**
How badly did I break it :)

### CODE REVIEW

**Comment by worktycho** [comment]
Are we really ready for c++14? Given that it is only supprted in gcc >4.9 and clang  >3.4 I think we need to wait until the compilers are more widely deployed.

**Comment by tigerw** [comment]
Well, are there any stats on how many people who wish to compile need to manually install a newer compiler anyway? Debian wheezy at least has clang 3.0 and gcc 4.7 (so it'll just be a version difference for them) and jessie meets both requirements. I guess the main problem is people who have already set up their build environment with something other than the latest having to upgrade.

For example, @madmaxoft, are you willing to update to VS 2015 or be unable to compile the server?

**Comment by madmaxoft** [comment]
I think VS2015 should be able to install on Win7, so I should be ok. Not sure about the compilers for the more exotic architectures, though - Android, RasPi etc.

Still, I think moving to C++14 would be too rushed at this point.

**Comment by tigerw** [comment]
Alright. So, do people think this will work?

File: src/Entities/Entity.cpp [hunk 1]
```diff
@@ -1451,7 +1450,7 @@ bool cEntity::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d
 	}
 
 	// Remove all links to the old world
-	SetWorldTravellingFrom(GetWorld());  // cChunk::Tick() handles entity removal
+	GetWorld()->GetChunkMap()->RemoveEntity(this);
```
Comment (worktycho) [inline src/Entities/Entity.cpp]: What thread does this get called from?

**Comment by worktycho** [comment]
What are the contention levels like on the locks?

File: src/Entities/Entity.cpp (hunk 1 above)
Comment (tigerw) [inline src/Entities/Entity.cpp]: I believe the world tick thread and whatever thread lets plugins run binding code.

**Comment by tigerw** [comment]
I'm not quite sure, wouldn't that depend on server load? How would you suggest I go about collecting this data?

**Comment by worktycho** [comment]
You'd need a profilier that can tell you it, but I'm not sure what the best one to do that is, as I've only used gprof which doesn't tell you that. However Intel VTune gets a lot of good coverage, claims to support lock profiling, and has an free linux version for open source projects.

**Comment by tigerw** [comment]
I did a concurrency profile with VS, and the locks in question didn't even show up. That seems to suggest that its effects are very small, but it might be that 1) I'm using the profiler wrong or 2) There's not enough load on the server.

Can you open VS profile files?

**Comment by worktycho** [comment]
Not at the moment. Did you connect to the server, or just start it?

**Comment by tigerw** [comment]
I connected and went through a portal a couple of times both ways.

**Comment by worktycho** [comment]
Ok, but I think we need to keep an eye on the locks for performance issues.

**Comment by tigerw** [comment]
So is that a yes?

**Comment by worktycho** [comment]
I still haven't properly reviewed it for race's. `cEntity *`'s can be held by other parts of the code at the moment so long as they only access it through the tick thread. This breaks the safety of that so you need to change all of those pointers to go through the lock.

**Comment by tigerw** [comment]
I believe all manipulation of retrieved entity pointers are done within the lock acquired before cChunk calls Tick() on each, so surely other threads can't delete an entity whilst they are being used?

**Comment by worktycho** [comment]
That is the the state before this patch. I assumed that you added the ability to manipulate entity pointers outside the cChunk lock, because otherwise, why do you need a separate lock for the list?

**Comment by tigerw** [comment]
Because I was thinking of moving across worlds, where one world thread adds an entity to another world thread, which needs synchronisation.

**Comment by madmaxoft** [comment]
Moving across worlds was done through buffers - the entity was moved from the original world to destination world's "add" buffer within original world's tick thread, and then the destination world's tick thread would move the entity from its "add" buffer into the world. It cannot be done in one step, because that would produce a lock dependency and thus a deadlock possibility.

**Comment by tigerw** [comment]
nooooooooooo

I'm debating two solutions:
1. To extend manual_lock_container with a queue. Inserts to the underlying container should go through a staged_push_back method. If a try_lock call on the mutex guarding the underlying container acquired it, the new object is pushed into the container. Otherwise, it is pushed onto the internal queue. This queue is processed by a separate thread, which, when alerted by a condition_variable set by the thread pushing onto the queue, calls std::lock on both mutexes (guarding the queue and underlying container), and when the locks are both successfully acquired, inserts the queue into the relevant data structure. The problem with this approach is that it adds a lot of complexity. The benefits are faster (but not guaranteed) pushing of entities into cChunk, and a reduction in work by the tick thread.
2. To changed manual_lock_container to a simple class encapsulating a container, a queue, and a lock guarding the queue, similar to what we have now, just as a class. The adder calls a method which pushes to the queue, and the tick thread calls a method which processes the queue.

I'm torn. What would you suggest? I guess there is really no way to avoid the deadlock of the current attempt?

**Comment by tigerw** [comment]
[Concurrency...](http://treasure.diylol.com/uploads/post/image/336787/resized_y-u-no-meme-generator-concurrency-y-u-so-hard-78948c.jpg)

**Comment by worktycho** [comment]
Its not difficult if you write pure functions, or keep state per thread. Its only concurrent access to shared state that's difficult to work with. :) But seriously, shared state is hard, so we should try and avoid sharing things where possible. (Queues are relatively simple, and we have a generic thread safe queue).

**Comment by tigerw** [comment]
Okay, I have another idea. Since problems are arising due to the need for other threads to concurrently access one world tick thread, it makes sense therefore to limit direct access to a cWorld, and any variables that are a member of it, to solely one thread. This could be partially enforced by not allowing cRoot to provide any world object. If one world wished to modify something in another, it must post a lambda (via cRoot) to the world in question (with variables capture by-value), which is then executed in the context of the tick thread.

Therefore, if everything is done right, I think this will eliminate the need completely for individual locks, except those which are directly exposed to concurrency.

This should work in theory, right? If so, does anyone know which threads currently interface with a world thread? Plugin hooks are executed synchronously, and network threads post to a queue, which is processed when clienthandles are ticked. So maybe the webadmin thread?

**Comment by worktycho** [comment]
No as per the following hole:

    postFunction([=worldPtr] (cWorld a_World) { /* copy from worldPtr to a_World */});

As far as I know, there is no way of using compile time interfaces to solve this problem in c++.

**Comment by tigerw** [comment]
What do you mean?

**Comment by worktycho** [comment]
as in whats to stop you posting a world pointer to another thread in a lambda.

**Comment by tigerw** [comment]
Whomever reviews the PR? :P
What stops me from forgetting to guard a new concurrently accessed vector with a lock? I can't really protect the code against the programmer, but if they try to pass a pointer in that fashion, at least it becomes obvious that something's wrong.

**Comment by worktycho** [comment]
And what about entity transfers? You need to ensure that thy are removed from on world before the post. We already review for thread safety, and c++ cannot provide the guarantees needed for static checks. There are scenarios where you don't have access to a cWorld object, but know you're on the right thread. So ScheduleTask should handle the cross thread execution issues.

**Comment by tigerw** [comment]
So for entity transfers, remove the entity entry from the old world and post the pointer onto the task queue of the new one, or, if cChunk owns smart pointers of entities, call a function which moves the ownership to the task lambda and deletes the entry afterwards.

Well, it would be a very interesting situation if one were processing on the world tick thread without any way to get a world pointer. If that is the case, one should really modify the arguments of whatever they're in to pass one.

**Comment by madmaxoft** [comment]
Also, we can't really lock the world access to the tick thread:
- it would create deadlocks with plugins executing world functions in the webadmin
- it would break the API for all plugins

**Comment by tigerw** [comment]
It would break plugins, since cRoot:GetWorld() can't exist, but this in theory is only currently used in the webadmin handler, where no world object is available, so changing it in every plugin is always do-able.

With the webadmin handler having to do work by passing a lambda to ScheduleTask, this will be guarded by a lock. But, since this won't depend on any other locks, where is the deadlock?

(forgive me if it's obvious, I'm still learning :P)

**Comment by madmaxoft** [comment]
It's used in most plugins, notably in the webadmin and console command handlers.

The deadlock is caused by the plugin code itself being guarded by a lock. Thus, a webadmin page goes to a Lua plugin, which tries to execute task on a world, but the world tick thread is waiting for the plugin to execute its hook.

Note that in order for most things to continue working, you need not only `ScheduleTask`-like function, but also `DoWith`-like function (immediate execution).

**Comment by tigerw** [comment]
I don't see how there could be a deadlock. A webadmin thread executing plugin code will not be given direct access to a cWorld, and therefore won't try to lock it. Instead, it should schedule a task (which depends on only one lock) via cRoot to execute within the context of the world thread. I just realised that I'm describing [this](http://mc-server.xoft.cz/LuaAPI/WebWorldThreads.html).

Meanwhile, plugins which are handling a hook will be doing so in the context of the world tick thread. By routing all extra-worldthread modifications through a schedule, all locks + the DoWith functions can be removed, since everything (entity creation/destruction, data (structure) changes, block changes, etc.) is serialised and confined to one thread.

In short, what I'm proposing is that:
* Only one thread changes a world directly
* Everything else is queued
* No more DoWith or providing world pointers
* No more locks in a world or anything that it contains, since everything is queued

Is it gonna work?

**Comment by madmaxoft** [comment]
It's not gonna work for plugins that need to interact with the world immediately (`DoWith`-style function). It would work only for postponed access (`ScheduleTask`-style function), but that would severely limit the abilities of plugins. For example webadmin functions need to return the entire HTML code immediately, they can't "schedule".a query and produce HTML based on that query.

**Comment by tigerw** [comment]
I think it will work for every place that's using DoWith now, since
* They are in the tick thread
* They have a world object
* So, they can access the world directly.

For the webadmin thread, according to [this](http://mc-server.xoft.cz/LuaAPI/WebWorldThreads.html) again, they aren't supposed to do anything related to unqueued access (ForEach, DoWith) because of what you said about deadlocks:
*  Webadmin: Plugin locked -> plugin incorrectly does a ForEach -> thread switch
*  World: Layers locked -> Tick -> hook handling -> plugin lock **blocked** -> thread switch
*  Webadmin: Layers lock **blocked**

**Comment by madmaxoft** [comment]
True, those deadlocks are a PITA. Before that could be done, though, we need a "resumable webadmin request" and "resumable command handler". That is, allow the handler to say "I have scheduled some work, the request / command will be finished later on from a callback" and the server will keep the request / command dangling until that callback finishes it.

**Comment by tigerw** [comment]
So does that mean the single thread world access is going to work? :D

I don't know about delaying the webadmin until a scheduled task finishes, but I'm just trying to argue that I think my proposals won't break anything.

**Comment by madmaxoft** [comment]
It could be made to work, yes. Yes, it will break things, but there is a workaround for those. The question is, do you want to invest that much work into something when you have very little expectancy of gains?

**Comment by tigerw** [comment]
Well, hopefully we can:

1. Eliminate all locks in cWorld (apart from the ScheduleTask bit) since nothing should be accessed out-of-thread
2. Eliminate all queuing, again, since other threads aren't trying to read/write concurrently.

I think even if there aren't many performance benefits, at least it makes the code clearer and smaller. And there won't be as many iterations of queues, so there's that.

**Comment by tigerw** [comment]
Why doesn't GitHub do automatic merges if the master is ahead but non-conflicted any more?

**Comment by madmaxoft** [comment]
Because we have required CI builds for the master branch.
I personally like that, because it prevents the tangled history. You should rebase before merging.

**Comment by worktycho** [comment]
Its an option I've turned on. I think it comes with not letting you merge unless CI has passed the build.

File: SetFlags.cmake [hunk 2]
```diff
@@ -62,6 +62,9 @@ macro(set_flags)
 		set(CMAKE_EXE_LINKER_FLAGS_RELEASE    "${CMAKE_EXE_LINKER_FLAGS_RELEASE}    /LTCG")
 		set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} /LTCG")
 		set(CMAKE_MODULE_LINKER_FLAGS_RELEASE "${CMAKE_MODULE_LINKER_FLAGS_RELEASE} /LTCG")
+
+		# Make build use Unicode:
+		add_definitions(-DUNICODE -D_UNICODE)
```
Comment (madmaxoft) [inline SetFlags.cmake]: Again: WHY? This has the potential to break many things and is not strictly needed for your goals.

File: src/Bindings/ManualBindings_World.cpp [hunk 3]
```diff
@@ -380,9 +378,9 @@ static int tolua_cWorld_QueueTask(lua_State * tolua_S)
 		return cManualBindings::lua_do_error(tolua_S, "Error in function call '#funcname#': Could not get function reference of parameter #1");
 	}
 
-	auto task = std::make_shared<cLuaWorldTask>(*Plugin, FnRef);
-	Plugin->AddResettable(task);
-	self->QueueTask(task);
+	auto ResettableTask = std::make_shared<cLuaWorldTask>(*Plugin, FnRef);
+	Plugin->AddResettable(ResettableTask);
+	self->QueueTask(std::bind(&cLuaWorldTask::Run, ResettableTask, std::placeholders::_1));
```
Comment (madmaxoft) [inline src/Bindings/ManualBindings_World.cpp]: I still prefer the old code style. It was clear and readable, unlike this mess of quadra-dots and a `std::bind` which is always a bad sign.

File: SetFlags.cmake (hunk 2 above)
Comment (worktycho) [inline SetFlags.cmake]: And if anything it should be in Globals.h along with all the other windows flags.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: I understand your concern, but wouldn't you say that one messy line is better than one class in World.h for every action which you want to perform with QueueTask? `std::bind` is the proper way to achieve this, despite that it doesn't look nice.

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: I agree that this weird hybrid of lambda's and strategy objects is a bad idea. Particularly as your taking a member function pointer and treating it as a function pointer.

What would be a much better way of doing this would be to take advantage of the fact that std::function will take anything that meets the Callable concept. So if you replace the Run method with an override of operator(), you should be able to construct a std::function from it.

File: SetFlags.cmake (hunk 2 above)
Comment (tigerw) [inline SetFlags.cmake]: Nothing appears to be untowardly broken, so why not? Unicode all the way, right?

I don't know enough to be able to respond to the Globals.h comment though, so okay, I guess.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: Really? I'll try that.

**Comment by tigerw** [comment]
you guys are such nitpickers :)

File: SetFlags.cmake (hunk 2 above)
Comment (worktycho) [inline SetFlags.cmake]: The big problem is that windows uses UCS-2(ish) whereas we use UTF-8. Now UTF-8 can be munged to windows codepage and back fairly safely. But UCS-2(ish) is really easy for UTF-8 systems to munge.

Comment (madmaxoft) [inline SetFlags.cmake]: Turning to unicode (especially on a single platform out of several) is a huge undertaking, it's not something you can just flip a switch and fix the few compiler warnings. The errors will be abundant and they will be subtle. Better not to do it in such a hurry.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (madmaxoft) [inline src/Bindings/ManualBindings_World.cpp]: You still have one class for each action that is to be scheduled, so there's no real improvement here - only that your construct doesn't inherit from a common predecessor.

File: src/OSSupport/Errors.cpp [hunk 4]
```diff
@@ -5,11 +5,11 @@
...
 AString GetOSErrorString( int a_ErrNo)
 {
-	char buffer[ 1024 ];
 	AString Out;
 
 	#ifdef _WIN32
 
+	wchar_t buffer[1024];
 	FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM, nullptr, a_ErrNo, 0, buffer, ARRAYCOUNT(buffer), nullptr);
 	Printf(Out, "%d: %s", a_ErrNo, buffer);
```
Comment (worktycho) [inline src/OSSupport/Errors.cpp]: This is broken because %s is for single byte encoding unit strings, not multi-byte encoding unit strings.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: Yes, lambdas still do create a class, but I was intending to convey that it is much less effort now that before, and World.h is that much smaller.

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: The improvement is that it lets us use lambda methods elsewhere.

File: SetFlags.cmake (hunk 2 above)
Comment (tigerw) [inline SetFlags.cmake]: I don't quite understan the munging bit. Why would "other systems" be affected by a flag and changes limited to Windows compilation?

Also, what do you mean by "we use UTF-8"? Do you mean the encoding of the source files?

Comment (worktycho) [inline SetFlags.cmake]: By we use utf-8, I'm referring to what our strings are encoded with. So all of our code expects a std::string to be in an encoding that uses single byte encoding units. In practice this means utf-8 or a windows code-page, but for a lot of string manipulation it doesn't actually matter which. 

UCS-2/UTF-16 is different enough that it can cause serious issues if it is passed to a function that expects a single byte encoding unit encoding. For example UCS-2/UTF-16 can have single byte embedded zero bytes in the middle of text, which breaks strcpy.

Comment (worktycho) [inline SetFlags.cmake]: And if you don't understand what an encoding unit is, you shouldn't be making this change.

Comment (tigerw) [inline SetFlags.cmake]: Well, I don't understand how std::function works either, doesn't mean I shouldn't experiment :)

Am I correct in thinking that, even under UNICODE, strings, both literal and dynamic, will only become wide if prefixed by `L` or given a type `wchar_t`? Because the server ran, printed messages, parsed recipes, etc. perfectly fine.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: I can't quite get std::fuction to work. Do you have any ideas that are one-liners?

It still seems to me the std::bind is the best way. The complaints thus far has been entirely of aesthetics, convention, and a fear of bind - do correct me if I'm wrong here. If it works, and it does excellently, why not use it?

File: SetFlags.cmake (hunk 2 above)
Comment (worktycho) [inline SetFlags.cmake]: Not understanding how std::function works is fine so long as you understand the API. encoding units is part of the API text encodings present.

And you are correct about the L prefix. The problem is that we don't use the windows API much, so testing this involves things like purposeful crashes.

File: src/Bindings/ManualBindings_World.cpp (hunk 3 above)
Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: Whats code you've written and the error? The signature of operator() should be `void operator()(cWorld & a_World);`

-----
My main worry with std::bind is that member function pointers are a mess of UB, and like varargs, often appear to work when UB.

Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: But I think the current state is correct? It's unlikely to change also.

Attempt:
`cpp
class cLuaWorldTask :
	public cPluginLua::cResettable
{
public:
	cLuaWorldTask(cPluginLua & a_Plugin, int a_FnRef) :
		cPluginLua::cResettable(a_Plugin),
		m_FnRef(a_FnRef)
	{
	}

	void operator()(cWorld & a_World)
	{
		cCSLock Lock(m_CSPlugin);
		if (m_Plugin != nullptr)
		{
			m_Plugin->Call(m_FnRef, &a_World);
		}
	}

protected:
	int m_FnRef;
};

Method1
{
	auto ResettableTask = std::make_shared<cLuaWorldTask>(*Plugin, FnRef);
	std::function<void(cWorld &)> CallableTask(std::placeholders::_1);
	CallableTask = ResettableTask;

	Plugin->AddResettable(ResettableTask);
	self->QueueTask(CallableTask);
}

Method2
{
	auto ResettableTask = std::make_shared<cLuaWorldTask>(*Plugin, FnRef);
	std::function<void(cWorld &)> CallableTask = ResettableTask;

	Plugin->AddResettable(ResettableTask);
	self->QueueTask(CallableTask);
}
```

`Severity	Code	Description	Project	File	Line
Error	C2893	Failed to specialize function template 'unknown-type std::invoke(_Callable &&,_Types &&...)' (compiling source file C:\Users\Tiger\Documents\GitHub\cuberite\src\Bindings\ManualBindings_World.cpp)	MCServer	C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\type_traits	1494
`

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: Ah. The problem is that you're passing a shared_ptr, not the cLuaWorldTask. And looking into it, fixing this is more complicated because of cResettable is a non-copyable type. A simple workaround is to use lambda for the bind, which is probably clearer.

Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: I can haz example?

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: `self->QueueTask([&ResettableTask] (cWorld & a_World) {ResettableTask->Run(a_World)});`
Or we can make cRessetable more supportive of the functional style, by adding:
```
template<class F>
void TryDoWithPlugin(F callback)
{
     cCSLock Lock(m_CSPlugin);
     if (m_Plugin != nullptr)
     {
          F(m_Plugin);
     }
}
```

Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: \*sigh\*

You and @madmaxoft sure? Bind was exactly meant for this and the llama introduces an extra intermediate class.

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: The point of TryDoWithPlugin would be to move the logic of cLuaWorldTask into the lambda. eg: 
```
self->QueueTask(
    [&Resettable, =FnRef] (cWorld & a_World)
    {
        Resettable->TryDoWithPlugin([&a_World, =FnRef] (cPlugin & a_Plugin)
        {
             a_Plugin->Call(FnRef, a_World);
        })
    });
```
But maybe this is too far down the road to callback hell.

Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: Yeah, probably. Are you like, absolutely sure about bind? I'm still not convinced.

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: This is the problem with functional code in imperative languages.
Maybe as SphinxC0re suggested on IRC what we really want is to be able to go full FRP with something like:
```
auto Task = self->QueueTask();
Task.flat_map([&Resettable](cWorld & a_World)
    {
         return Resettable.TryDoWithPlugin();
    })
    .zip(Task)
    .then([=FnRef](std::tuple<cPlugin, cWorld & a_args)
    {
         (a_args.first)->Call(FnRef, a_args.second);
    });

Comment (worktycho) [inline src/Bindings/ManualBindings_World.cpp]: Or another option:
Resurface locked_ptr, which should be simple to implement with a custom deleter and unique_ptr.
Then we can get rid of the inner callback in the callback hell one.

Comment (tigerw) [inline src/Bindings/ManualBindings_World.cpp]: Perhaps we are over-engineering this somewhat. Could we hold the idea for another time and just use std::bind? It seems the easiest and most performant option.

File: SetFlags.cmake (hunk 2 above)
Comment (tigerw) [inline SetFlags.cmake]: I can confirm that a simulated crash writes a core dump file correctly. I can further confirm that on changing the format string in Errors.cpp, errors are printed correctly.

This good?

**Comment by tigerw** [comment]
No farther communication, assumed acceptable?

**Comment by worktycho** [comment]
I was hoping to get @madmaxoft 's opinion on the bind alternatives.

File: src/World.cpp [hunk 5]
```diff
@@ -962,55 +961,39 @@ void cWorld::TickMobs(std::chrono::milliseconds a_Dt)
...
-			else
+			return;
+		}
+
+		// Partition everything to be executed by returning false to move to end of list if time reached
+		auto MoveBeginIterator = std::partition(m_Tasks.begin(), m_Tasks.end(), [this](const decltype(m_Tasks)::value_type & a_Task)
 			{
-				// All the eligible tasks have been moved, bail out now
-				break;
+				if (a_Task.first <= std::chrono::duration_cast<cTickTimeLong>(m_WorldAge).count())
```
Comment (worktycho) [inline src/World.cpp]: Should Tasks execute after x Ticks time or after x Ticks have been executed? The current code is executed after x Ticks time. It can cause subtle differences when lagging and I suspect that the current cod is not the expected behavior.

Comment (tigerw) [inline src/World.cpp]: Indeed, should have thought about that more thoroughly. Many thanks!

Comment (madmaxoft) [inline src/World.cpp]: I believe we should aim to process all game logic in logical ticks instead of wall clock time
If the ticks take too long, the server could limit the actions that are processed somehow, but delaying tasks is not a good way of doing so, especially if such tasks belong to core logic.

**Comment by madmaxoft** [comment]
As for the `bind` alternatives: I personally really dislike the "full FRP" style. It's too wordy and yet it doesn't easily say what it's doing.

So far I like this code the most:
```cpp
self->QueueTask(
    [&Resettable, =FnRef] (cWorld & a_World)
    {
        Resettable->TryDoWithPlugin([&a_World, =FnRef] (cPlugin & a_Plugin)
        {
             a_Plugin->Call(FnRef, a_World);
        })
    });
```
Since the call into a plugin would be one of the more common operations that will be queued / scheduled, it would make sense to get rid of the inner lambda by writing a templated `cResettable::Call(...)` function.

File: src/World.cpp (hunk 5 above)
Comment (worktycho) [inline src/World.cpp]: The reason the code is like this is we don't have a good measure of world age in ticks at the moment. m_WorldAge is the only measure we have and its wall clock age.

**Comment by worktycho** [comment]
Something along the following lines would work:
```
template <class... Args>
void cResetable::CallOnWorld(cWorld & a_World, Args&&... a_args)
{
	a_world.QueueTask(
		[=this, =FnRef, &a_args...] (cWorld & a_World)
		{
			cCSLock Lock(m_CSPlugin);
			if (m_Plugin != nullptr)
			{
				m_Plugin->Call(m_FnRef, &a_World, std::forward(a_args)...);
			}
		});
}
```
By moving the method into cResetable, we remove the need for TryDoWithPlugin, meaning we only need one lambda.

**Comment by madmaxoft** [comment]
Okay, that looks reasonable.

**Comment by tigerw** [comment]
We unfortunately have a bit of an impasse here because this
```cpp
template <class... Args>
void cResetable::CallOnWorld(cWorld & a_World, Args&&... a_args)
{
	a_world.QueueTask(
		[=this, =FnRef, &a_args...] (cWorld & a_World)
		{
			cCSLock Lock(m_CSPlugin);
			if (m_Plugin != nullptr)
			{
				m_Plugin->Call(m_FnRef, &a_World, std::forward(a_args)...);
			}
		});
}
```
was not a change I had in mind to make when I decided upon this
```cpp
std::bind(&cLuaWorldTask::Run, ResettableTask, std::placeholders::_1)
```

I still don't have a reason why bind shouldn't be used.

**Comment by worktycho** [comment]
@tigerw Sometimes (actually quite frequently in this code base) you find that one refactor needs another to work well. I sometime find myself so deep in dependent refactors that I forget what the original refactor was. For example #2488 was required for a protocol refactor I'm working on, which is required for #2477.
I personally think xofts suggestion is the best way of handling this. If you don't think it should go in this PR, I'll put it in another one. We can then rebase this PR on to it.

**Comment by tigerw** [comment]
I see what you're saying, but in this case, *why* are the suggestions necessarily better?

**Comment by worktycho** [comment]
I'm looking at the code again, and fundamentally the problem is that cRessettable's interface doesn't work well with the functional style. In the functional style you're going for cLuaWorldTask shouldn't exist. So the functional way would be to go with my TryDoWithPlugin method.

It doesn't look to me like CallOnWorld belongs in cResettable. I'm wondering if a better way to handle this is to refactor cRessable into a non-owning atomic smart_ptr. I'm just going to look at the concurrency TS to see what atomic_weak_ptr is going to look like.

**Comment by worktycho** [comment]
Atomic_shared_ptr did not give me any insperation, and I'm thinking the locked pointer option is the best. It's the only one I haven't written up yet, so I'll try and remember to do that tomorrow.

**Comment by tigerw** [comment]
Do I detect reservations? :P

The thought of introducing atomicity or further locking into this, or even, as you say, contaminating cResettable with task scheduling code, is not filling me with confidence of the practicality of any of this.

If cLuaWorldTask isn't functional, then fine. Should we really stick vigorously to convention and ideas about what code *should* be, at the expense of clarity?

**Comment by worktycho** [comment]
The reason I care about cLuaWorldTask not being functional, is that the whole problem here is that you're trying to fit a strategy object into a space for function. If we can fix that the whole complication with bind disappears.

**Comment by tigerw** [comment]
Is fitting a strategy object into the space for a function bad? And bind isn't really that complicated.

**Comment by worktycho** [comment]
its messy, and the reason you're doing it is a pattern that comes up frequently, so its worth sorting out.

**Comment by tigerw** [comment]
Well, alright.

So merge first and do whatever you want later?

**Comment by worktycho** [comment]
I think we should clean the code up as it comes in, because otherwise cleanup PR's don't happen that much.

**Comment by tigerw** [comment]
But [this guy (point 3)](http://hintjens.com/blog:95) says that open source projects should merge first and ask questions later!

**Comment by worktycho** [comment]
Yes, and that was how we ended up with god objects in the first place.

**Comment by tigerw** [comment]
You could also argue that a strict approvals process is how there are 21 open pull requests, most abandoned, and that this isn't exactly beneficial to a open source project aiming to replicate a server for a rapidly-evolving game, as can be seen by our very erratic new changes/features rate.

Please let's merge if there isn't anything really serious?

**Comment by worktycho** [comment]
And Neither is that fact that we have a number of crash bugs originating in race conditions. Most of which are really hard to find due to confusing control flow or ownership. Its a matter of which is more important: feature parity, or the ability to run for a length of time without crashing.

**Comment by tigerw** [comment]
Well, this PR and subsequent others are planning to simplify and hopefully fix all of that :)

**Comment by worktycho** [comment]
Good luck sorting that. Also at this point I think I'd merge this if you replaced std::bind with a lambda that did the same thing, as personally I think its clearer. Otherwise, I think we need @madmaxoft's opinion on what to do as our discussion is just degenerating into an argument about the importance of code quality.

**Comment by madmaxoft** [comment]
Get rid of the Unicode change, I'm not letting that in without serious testing.

As for the bind, I think I can live with it, despite grinding my teeth about it. The code could use a bit more commenting, though.

And change the commit message header to something more descriptive.

**Comment by SphinxC0re** [comment]
:pray: thanks for that
